app/uk/gov/hmrc/pillar2submissionapi/controllers/actions/AuthenticatedIdentifierAction.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.pillar2submissionapi.controllers.actions
+
+import com.google.inject.{Inject, Singleton}
+import play.api.Logging
+import play.api.mvc.Results._
+import play.api.mvc._
+import uk.gov.hmrc.auth.core.AffinityGroup.{Agent, Organisation}
+import uk.gov.hmrc.auth.core.AuthProvider.GovernmentGateway
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.pillar2submissionapi.models.requests.IdentifierRequest
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class AuthenticatedIdentifierAction @Inject() (
+  override val authConnector:    AuthConnector,
+  val parser:                    BodyParsers.Default
+)(implicit val executionContext: ExecutionContext)
+    extends IdentifierAction
+    with AuthorisedFunctions
+    with Logging {
+
+  override def refine[A](request: Request[A]): Future[Either[Result, IdentifierRequest[A]]] = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    authorised(AuthProviders(GovernmentGateway) and ConfidenceLevel.L50)
+      .retrieve(
+        Retrievals.internalId and Retrievals.groupIdentifier
+          and Retrievals.clientId and Retrievals.affinityGroup
+          and Retrievals.credentialRole and Retrievals.credentials
+      ) {
+
+        case Some(internalId) ~ Some(groupId) ~ Some(clientId) ~ Some(Organisation | Agent) ~ Some(User) ~ credentials =>
+          Future.successful(
+            Right(
+              IdentifierRequest(
+                request,
+                internalId,
+                Some(groupId),
+                clientPillar2Id = clientId,
```

Comment:
Is this client Id actually the pillar2 id? we don't need to fetch from enrolments?

Reply:
If true then it brings up a headache in the case that we get provided a pillar2 id in the request but after retrieving enrolments they do not have a pillar2 id. A error case we'll have to cover